### PR TITLE
Create JIRA client bundle

### PR DESCRIPTION
**Pull request by arothuis**
[Ticket 104424786](https://www.pivotaltracker.com/story/show/104424786)

### CODE REVIEW

**Comment by arothuis** [comment]
@rjkip @DRvanR

File: src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    private $client;
+
+    /**
+     * @param HttpClientInterface $client
+     * @param LoggerInterface $logger
+     */
+    public function __construct(HttpClientInterface $client, LoggerInterface $logger)
+    {
+        $this->client = $client;
+        $this->logger = $logger;
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php]: You inject a logger, yet you don't log anything.

File: src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * @param $method
+     * @param $url
+     * @param array $data
+     * @param $endpoint
+     * @param AuthenticationInterface $credential
+     * @return \Psr\Http\Message\StreamInterface
+     */
+    // @codingStandardsIgnoreStart (arguments with default values must be at the end; sadly, our JiraClient does not agree)
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php]: Noice.

File: src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param $method
+     * @param $url
+     * @param array $data
+     * @param $endpoint
+     * @param AuthenticationInterface $credential
+     * @return \Psr\Http\Message\StreamInterface
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php]: Nit pick: import StreamInterface

Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php]: Scratch that, just remove the PHPDoc, you only implement the interface and there's nothing more to document.

File: src/Surfnet/JiraApiClientBundle/Resources/config/services.yml [hunk 4]
```diff
@@ -1,4 +1,12 @@
 services:
-#    jira_api_client.example:
-#        class: Surfnet\JiraApiClientBundle\Example
-#        arguments: [@service_id, "plain_value", %parameter%]
+jira_api_client.guzzle:
+    class: GuzzleHttp\Client
+    arguments:
+      base_uri: @jira_api_client.api_url
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/Resources/config/services.yml]: This should be a parameter lookup, not a service.

File: src/Surfnet/JiraApiClientBundle/Resources/config/services.yml [hunk 5]
```diff
@@ -1,4 +1,12 @@
 services:
-#    jira_api_client.example:
-#        class: Surfnet\JiraApiClientBundle\Example
-#        arguments: [@service_id, "plain_value", %parameter%]
+jira_api_client.guzzle:
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/Resources/config/services.yml]: To define services, place them under `services:` (ie. indent once).

File: src/Surfnet/JiraApiClientBundle/Resources/config/services.yml [hunk 6]
```diff
@@ -1,4 +1,12 @@
 services:
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/Resources/config/services.yml]: Is there a service definition for the actual API client?

File: src/Surfnet/JiraApiClientBundle/Tests/DataProvider/DataProvider.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+trait DataProvider
+{
+    public function notNonStringScalarProvider()
+    {
+        return [
+            'boolean'      => [true],
+            'integer'      => [1],
+            'float'        => [1.2],
+        ];
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/Tests/DataProvider/DataProvider.php]: What about nulls, arrays, objects and -- however unlikely, but not hard to test -- resources?

File: src/Surfnet/JiraApiClientBundle/Tests/DataProvider/DataProvider.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\JiraApiClientBundle\Tests\DataProvider;
+
+trait DataProvider
+{
+    public function notNonStringScalarProvider()
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/Tests/DataProvider/DataProvider.php]: Should this be `nonStringScalarProvider`?

File: src/Surfnet/JiraApiClientBundle/DependencyInjection/JiraApiClientExtension.php [hunk 9]
```diff
@@ -31,9 +31,15 @@ class JiraApiClientExtension extends Extension
...
+        $config = $this->processConfiguration($configuration, $configs);
 
         $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
         $loader->load('services.yml');
+
+        $container->setParameter('jira_api_client.host', $config['host']);
+        $container->setParameter('jira_api_client.username', $config['username']);
+        $container->setParameter('jira_api_client.password', $config['password']);
+        $container->setParameter('jira_api_client.project_key', $config['project_key']);
+        $container->setParameter('jira_api_client.default_assignee', $config['default_assignee']);
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/DependencyInjection/JiraApiClientExtension.php]: @DRvanR What was your aversion against parameters (vs. definition argument replacement)?

File: src/Surfnet/JiraApiClientBundle/Tests/DependencyInjection/ConfigurationTest.php [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+    /**
+     * @test
+     * @group JiraApiClientBundle
+     * @dataProvider notNonStringScalarProvider
+     */
+    public function host_cannot_be_other_than_string($value)
+    {
+        $config = $this->validConfig;
+        $config["host"] = $value;
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/Tests/DependencyInjection/ConfigurationTest.php]: Nice, easy to understand :+1:

File: src/Surfnet/JiraApiClientBundle/Tests/DependencyInjection/ConfigurationTest.php [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        $this->assertConfigurationIsInvalid([$config], "The default assignee name should be a string");
+    }
+
+    /**
+     * Return the instance of ConfigurationInterface that should be used by the
+     * Configuration-specific assertions in this test-case
+     *
+     * @return \Symfony\Component\Config\Definition\ConfigurationInterface
+     */
+    protected function getConfiguration()
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/Tests/DependencyInjection/ConfigurationTest.php]: Please remove unnecessary PHPDoc.

File: src/Surfnet/JiraApiClientBundle/Resources/config/services.yml (hunk 6 above)
Comment (arothuis) [inline src/Surfnet/JiraApiClientBundle/Resources/config/services.yml]: There is now!

File: src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php (hunk 1 above)
Comment (arothuis) [inline src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php]: We'll do it later. This PR is more about the general wiring than the actual implementation of how the client (including the logging) is going to function.

Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php]: :+1:

**Comment by arothuis** [comment]
@rjkip :poultry_leg:

File: src/Surfnet/JiraApiClientBundle/DependencyInjection/Configuration.php [hunk 12]
```diff
@@ -21,15 +21,71 @@
...
+
+        $rootNode
+            ->children()
+                ->scalarNode('base_url')
+                    ->info('The base URL of the JIRA API host')
+                    ->isRequired()
+                    ->cannotBeEmpty()
+                    ->validate()
+                        ->ifTrue(function ($host) {
+                            return !is_string($host);
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/DependencyInjection/Configuration.php]: Is it a valid URL (`filter_var`)? Make sure it doesn't end in a forward slash (API calls [start with a forward slash](https://github.com/chobie/jira-api-restclient/blob/661e028fa3e96451202307259c4695d25df1a393/src/Jira/Api.php#L121)). Rename `$host` to `$baseUrl` while you're there :wink:

File: src/Surfnet/JiraApiClientBundle/DependencyInjection/Configuration.php [hunk 13]
```diff
@@ -21,15 +21,71 @@
...
+                        ->thenInvalid('The JIRA API username should be a string')
+                    ->end()
+                ->end()
+                ->scalarNode('password')
+                    ->info('The password of the user that will act as a reporter in JIRA')
+                    ->isRequired()
+                    ->cannotBeEmpty()
+                    ->validate()
+                        ->ifTrue(function ($password) {
+                        return !is_string($password);
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/DependencyInjection/Configuration.php]: Nit-pick: indentation.

File: src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php (hunk 1 above)
Comment (DRvanR) [inline src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php]: The property itself has not been defined though :wink:

File: src/Surfnet/JiraApiClientBundle/Resources/config/services.yml [hunk 14]
```diff
@@ -1,4 +1,24 @@
...
+    class: Surfnet\Conext\JiraApiClientBundle\HttpClient
+    arguments:
+      - @jira_api_client.guzzle
+      - @logger
+
+  jira_api_client.authentication:
+    class: Jira_Api_Authentication_Basic
+    arguments:
+      - @jira_api_client.username
+      - @jira_api_client.password
```
Comment (DRvanR) [inline src/Surfnet/JiraApiClientBundle/Resources/config/services.yml]: if it are parameters, shouldn't they it be `%jira_api_client.*` rather than `@...`?

**Comment by DRvanR** [comment]
Seem to be missing the actual configuration, the kernel update and the parameters.yml.dist update?

File: src/Surfnet/JiraApiClientBundle/DependencyInjection/JiraApiClientExtension.php (hunk 9 above)
Comment (DRvanR) [inline src/Surfnet/JiraApiClientBundle/DependencyInjection/JiraApiClientExtension.php]: Yes, parameters that do not need to be in the global scope for retrieval (e.g. an admin mail address may be reused by different services, thus useful to have available) should not be in the global scope. `%kernel.debug%` is reused throughout, really nice to have available globally. The jira credentials not so much :wink: 

I.m.o. as a bundle you provide additional functionality through your API, in this case publicly defined services with a known interface (at least, once we get round to the actual implementation :wink:). The jira information is local to this bundle only, and is only used to configure internal services and not relevant to anything outside the scope of this bundle. Therefor they should not available through the container for others to retrieve.

Comment (arothuis) [inline src/Surfnet/JiraApiClientBundle/DependencyInjection/JiraApiClientExtension.php]: Makes sense. Will fix this and others tomorrow.
Op 5 okt. 2015 18:27 schreef "Daan van Renterghem" <notifications@github.com

**Comment by arothuis** [comment]
@DRvanR @rjkip

File: src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    public function __construct(
+        HttpClientInterface $client,
+        LoggerInterface $logger,
+        $projectId,
+        $defaultAssignee
+    ) {
+        $this->client = $client;
+        $this->logger = $logger;
+        $this->projectId = $projectId;
+        $this->defaultAssignee = $defaultAssignee;
```
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/HttpClient/HttpClient.php]: As discussed, please remove `projectId` and `defaultAssignee`. These should be handled at service level.

File: src/Surfnet/JiraApiClientBundle/Tests/DataProvider/DataProvider.php (hunk 7 above)
Comment (rjkip) [inline src/Surfnet/JiraApiClientBundle/Tests/DataProvider/DataProvider.php]: Did we determine that `null` was also a valid value for `scalarNode`? Please add if that is the case.

**Comment by rjkip** [comment]
Bar the failed test and missing scalar config `null` test, :+1: from me.

**Comment by DRvanR** [comment]
Agree witth @rjkip

**Comment by arothuis** [comment]
1. Fixed the test. 
2. Null already fails the cannotBeEmpty() test, so it is not part of any custom validation.

**Pull request merged**